File: codes/csharp/chapter_greedy/max_capacity.cs
/**
* File: max_capacity.cs
* Created Time: 2023-07-21
* Author: hpstory (hpstory1024@163.com)
*/

namespace hello_algo.chapter_greedy;

public class max_capacity {
    /* 最大容量：贪心 */
    int MaxCapacity(int[] ht) {
        // 初始化 i, j，使其分列数组两端
        int i = 0, j = ht.Length - 1;
        // 初始最大容量为 0
        int res = 0;
        // 循环贪心选择，直至两板相遇
        while (i < j) {
            // 更新最大容量
            int cap = Math.Min(ht[i], ht[j]) * (j - i);
            res = Math.Max(res, cap);
            // 向内移动短板
            if (ht[i] < ht[j]) {
                i++;
            } else {
                j--;
            }
        }
        return res;
    }

    [Test]
    public void Test() {
        int[] ht = [3, 8, 5, 2, 7, 7, 3, 4];

        // 贪心算法
        int res = MaxCapacity(ht);
        Console.WriteLine("最大容量为 " + res);
    }
}

File: codes/scala/chapter_greedy/max_capacity.scala

/**
 * File: maxCapacity.scala
 * Author: WangBlue (wangjiping596@gmail.com)
 * Create: 2024-08-18 11:48
 */


object max_capacity {

  /* 最大容量：贪心 */
  def maxCapacity(ht: Array[Int]) = {
    // 初始化 i, j，使其分列数组两端
    var i = 0
    var j = ht.length - 1
    // 初始最大容量为 0
    var res = 0
    // 循环贪心选择，直至两板相遇
    while (i < j) {
      // 更新最大容量
      val cap = Math.min(ht(i), ht(j)) * (j - i)
      res = Math.max(res, cap)
      // 向内移动短板
      if (ht(i) < ht(j)) i += 1
      else j -= 1
    }
    res
  }

  def main(args: Array[String]): Unit = {
    val ht = Array(3, 8, 5, 2, 7, 7, 3, 4)

    // 贪心算法// 贪心算法
    val res = maxCapacity(ht)
    println("最大容量为 " + res)

  }

}
